fix: add some missing workspace updates

code-asher opened the pull request:
Fixes https://github.com/coder/coder/issues/7186.  Probably.  I was not able to reproduce with a template that had invalid credentials (I used the Heroku template) but I was able to reproduce with the Docker template by deleting my volumes (well I restarted my dev workspace which killed the volumes).

There were two issues:

1. Sometimes a failed job does not have the workspace build type so no workspace update was published.  Fix is to add that type to all workspace build failures (or at least all the ones I found).
2. Sometimes a failed job does not have any state so again no workspace update was published as that is skipped when there is no state.  Fix is to publish an update even when there is no state.

code-asher committed:
Standardize on workspace channel function

code-asher committed:
Add some missing workspace updates

There are some failure cases where we do not set the type as a workspace
build which causes the workspace update to never be published.

code-asher commented on coderd/provisionerdserver/provisionerdserver.go:
```diff
@@ -757,9 +757,6 @@ func (server *Server) FailJob(ctx context.Context, failJob *proto.FailedJob) (*p
 
 	switch jobType := failJob.Type.(type) {
 	case *proto.FailedJob_WorkspaceBuild_:
-		if jobType.WorkspaceBuild.State == nil {
-			break
-		}
```
I am not sure what this was guarding against so maybe removing it has some unintended consequences...

code-asher replied:
I added it back but only gated the update behind it since that appears specifically for updating the state and I suppose we might not want to do that when no state was included but we still want the publish.

code-asher commented on provisionerd/runner/runner.go:
```diff
@@ -1116,6 +1116,7 @@ func (r *Runner) failedJobf(format string, args ...interface{}) *proto.FailedJob
 		JobId:     r.job.JobId,
 		Error:     message,
 		ErrorCode: code,
+		Type:      &proto.FailedJob_WorkspaceBuild_{},
```
I was able to trigger this by using the Docker template, creating a workspace, deleting my Docker volumes, then trying to start the workspace again.

It does not trigger through the unit test.  Not sure how to make it take this code path.

code-asher commented on coderd/workspaces_test.go:
```diff
@@ -1733,13 +1733,37 @@ func TestWorkspaceWatcher(t *testing.T) {
 
-	cancel()
+	// Build with the new template; should end up with a failure state.
+	_ = coderdtest.CreateWorkspaceBuild(t, client, workspace, database.WorkspaceTransitionStart, func(req *codersdk.CreateWorkspaceBuildRequest) {
```
For some reason creating this build fails the test, somewhere I think there is an assertion that the provisioner has no errors or something.

code-asher replied:
Turns out it was the logger failing the test due to an error log when testing the failure state.

This implies to me that the logger's opinion is either that testing failures is an anti-pattern (which seems wrong) or that a provisioner job failing does not qualify to be logged as an error (could maybe change it to a warning).

For now I added a bool to disable this behavior rather than globally disable it or change the log level.

code-asher commented on provisionerd/runner/runner.go:
```diff
@@ -1102,6 +1102,12 @@ func (r *Runner) runWorkspaceBuild(ctx context.Context) (*proto.CompletedJob, *p
 	}, nil
 }
 
+func (r *Runner) failedWorkspaceBuildf(format string, args ...interface{}) *proto.FailedJob {
```
I feel like there must be a better way to do this...maybe `failedJobf().withWorkspaceBuildType()` or something but `proto.FailedJob` is generated so not sure how to add that.

kylecarbs replied:
Meh, I think this is fine

kylecarbs commented on coderd/coderdtest/coderdtest.go:
```diff
@@ -134,6 +134,10 @@ type Options struct {
 
 	SwaggerEndpoint bool
+
+	// LoggerIgnoreErrors causes the test logger to not fatal the test on Fatal
+	// and not error the test on Error or Critical.
+	LoggerIgnoreErrors bool
```
I'm not sure we should make this an option... if an error isn't a failure, we should change the log.

If I'm misunderstanding, please link the log in question so I can take a closer look 🕵️🕵️🕵️

code-asher replied:
Yeah I was quite conflicted about this.  It implies to me that the logger's opinion is either that testing failures is an anti-pattern or that a workspace build job failing does not qualify to be logged as an error.

I believe it was this log here: https://github.com/coder/coder/blob/467b0a10046337570b9a7d21f6e4609c79ecf809/provisionerd/runner/runner.go#L929-L932

Maybe this should only be warning level?

kylecarbs replied:
Ahh, yes. I'd say warning level is fine!

code-asher committed:
Make build failures warnings

Otherwise the associated test fails due to the logger fataling on
error messages.
